lib/options_validator.js

```diff
@@ -167,4 +167,85 @@ function invalidate(validationLevel, message, logger) {
   }
 }
 
-module.exports = { assertArity, createValidationFunction };
+/**
+ * This is a class to build an operation by checking its arity and validating its options.
+ */
+class OperationBuilder {
+  constructor() {
+    this.requiredArity = -1;
+    this.validationSchema = {};
+  }
+
+  /**
+   * Set an operation's required arity.
+   *
+   * @method
+   * @param {number} requiredArity The operation to validate.
+   * @return {OperationBuilder} An operation builder with a set requiredArity.
+   */
+  arity(requiredArity) {
+    this.requiredArity = requiredArity;
+    return this;
+  }
+
+  /**
+   * Set an OperationBuilder's validationSchema.
+   *
+   * @method
+   * @param {object} validationSchema The schema against which to validate options.
+   * @return {OperationBuilder} An operation builder that will validate options.
+   */
+  options(validationSchema) {
+    this.validationSchema = validationSchema;
+    return this;
+  }
+
+  /**
+   * Build an operation by checking its arity and validating its options.
+   *
+   * @method
+   * @param {function} operationToBuild The operation to validate.
+   * @return {function} The operation with validated options.
+   */
+  build(operationToBuild, overrideOptions, validationOptions) {
+    const operationBuilder = this;
+
+    function buildOperation() {
+      assertArity(arguments, operationBuilder.requiredArity);
+
+      const args = Array.prototype.slice.call(arguments);
+
+      // validate options
+      const optionsIndex =
+        typeof args[args.length - 1] === 'function' ? args.length - 2 : args.length - 1;
+
+      const verifiedOptions = validationFunction(
+        operationBuilder.validationSchema,
+        args[optionsIndex],
+        overrideOptions,
+        validationOptions
+      );
+
+      args[optionsIndex] = verifiedOptions;
+
+      // call original function with validated options
+      return operationToBuild.apply(this, args);
+    }
+
+    return buildOperation;
+  }
+}
+
+/**
+ * Build an OperationBuilder given an operation's required arity.
+ *
+ * @method
+ * @param {number} requiredArity The required arity of the operation to build.
+ * @return {OperationBuilder} An operation builder that will validate options.
+ */
+function arity(requiredArity) {
```

Comment:
does it make more sense to just have `arityZero`, `arityOne`, `arityTwo`, rather than potentially letting someone enter an invalid value?

Reply:
I don't think a user should be using any of this so we're the only ones who could enter an invalid value. I do think that `arityZero`, `arityOne`, `arityTwo` could protect us from ourselves so I'll go ahead and add that.